### DISCUSSION TITLE
Pitchfork Worker Killer

### DISCUSSION THREAD

**Discussion by jpcamara** [Shopify/pitchfork#75] [Ideas]
In unicorn deployments I currently use https://github.com/kzk/unicorn-worker-killer. I'm curious to start testing real usage against Pitchfork, but would like to still maintain the ability to prune workers based on these same characteristics.

Do you see any issues with a basic fork of that gem which renames the Unicorn pieces to Pitchfork? Or is there anything more fundamental that would be a mismatch between the Pitchfork approach and killing processes based on memory and request count statistics?

**Comment by casperisfine** [comment]
> Do you see any issues with a basic fork of that gem which renames the Unicorn pieces to Pitchfork?

Some very minor things, like the `process_client` method now takes a couple extra params. But more importantly, `Pitchfork` now expose an [`after_request_complete`](https://github.com/Shopify/pitchfork/blob/master/docs/CONFIGURATION.md#after_request_complete) callback which allows to implement this pattern without resorting to monkey patching.

So it would require very little adaptation, but would be cleaner.

Also for `MaxRequests` specifically, `Worker#requests_count` is exposed, so the code could be simplified too.

As for `Oom`, it uses `RSS` which is a very misleading metric, even more so with reforking enabled, so ideally it would be changed to use PSS instead, which is provided by `Pitchfork::MemInfo`: https://github.com/Shopify/pitchfork/blob/891bc8ee7d547bcc29948f2aaea1892e41e09c4f/lib/pitchfork/mem_info.rb#L4

**Comment by jpcamara** [comment]
Nice, thanks for the breakdown!
